File: 341/341_version2.py
class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.stack = []
        # initialize the stack 
        index = len(nestedList) - 1
        while index >= 0:
            self.stack.append(nestedList[index])
            index-=1
        

    def next(self):
        """
        :rtype: int
        """
        return self.stack.pop().getInteger()

    def hasNext(self):
        """
        :rtype: bool
        """
        while self.stack:
            top = self.stack[-1]
            if top.isInteger(): return True
            self.stack.pop()
            restList = top.getList()
            index = len(restList) - 1
            while index >=0:
                self.stack.append(restList[index])
                index-=1
        return False

File: 341/341_version2.java
public class NestedIterator implements Iterator<Integer> {
    private Stack<NestedInteger> stack = new Stack<>();
    public NestedIterator(List<NestedInteger> nestedList) {
        for(int i = nestedList.size()-1;i>=0;i--) stack.push(nestedList.get(i)); // push all elements into stack
    }

    @Override
    public Integer next() {
        return stack.pop().getInteger();
    }

    @Override
    public boolean hasNext() {
        while(!stack.empty()){
            NestedInteger top = stack.peek();
            if(top.isInteger()) return true;
            stack.pop();// pop out the top element: which is a List<NestedInteger>
            for(int i = top.getList().size()-1;i>=0;i--) stack.push(top.getList().get(i));
        }
        return false;
        
    }
    
}
